### PR TITLE
New Provider: OpsGenie

### LINKED ISSUE #9557: Missing OpsGenie Support
Hi there,

I noticed that PagerDuty support was just implemented. It would be nice to also have support for OpsGenie. Thanks!
### Terraform Version

Terraform v0.7.7
### Expected Behavior

Support for OpsGenie
### Actual Behavior

Not yet.
### References
- WebAPI https://www.opsgenie.com/docs/dashboard/web-api
- GoSDK https://github.com/opsgenie/opsgenie-go-sdk

### PR DESCRIPTION
**Pull request by tombuildsstuff**
This is my first time writing a new provider.. so I might be missing something, but here's what I'm aware of:

 - [x] Add support for the new Provider
  - [x] Provider configured via an API Key
  - [x] Documentation (Sidebar/Providers page)
    - [x] Sidebar
    - [x] Providers
    - [x] Instructions for how to generate the API Key
 - [x] Import the Vendored SDK
 - [x] Data Sources
   - [x] Users
      - [x] Read / Tests
      - [x] Documentation
 - [x] Resources
    - [x] Teams
      - [x] CRUD
      - [x] Import
      - [x] Field Validation
      - [x] Tests
      - [x] Documentation
    - [x] Users
      - [x] CRUD
      - [x] Import
      - [x] Field Validation
      - [x] Tests
      - [x] Documentation

In a follow up PR:
 - Optional Fields:
  - Teams: `Description` - [waiting on an SDK change](https://github.com/opsgenie/opsgenie-go-sdk/pull/8)
  - Users: `SkypeUsername` - [waiting on an SDK Change](https://github.com/opsgenie/opsgenie-go-sdk/pull/10)

I'll add support for Contacts in a follow up PR - but that also needs an SDK change..

Fixes #9557

### CODE REVIEW

File: website/source/docs/providers/opsgenie/index.html.markdown [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+```
+
+## Configuration Reference
+
+The following arguments are supported:
+
+* `api_key` - (Required) The API Key for the OpsGenie Integration. If omitted, the
+  `OPSGENIE_API_KEY` environment variable is used.
+
+TODO: more details on setting up an integration (including access types required)
```
Comment (tombuildsstuff) [inline website/source/docs/providers/opsgenie/index.html.markdown]: need to do this

**Comment by tombuildsstuff** [comment]
Tests pass:

```
$ TF_ACC=1 go test ./builtin/providers/opsgenie -v -run TestAcc -timeout 120m
=== RUN   TestAccDataSourceOpsGenieUser_Basic
--- PASS: TestAccDataSourceOpsGenieUser_Basic (6.92s)
=== RUN   TestAccOpsGenieTeam_importBasic
--- PASS: TestAccOpsGenieTeam_importBasic (4.92s)
=== RUN   TestAccOpsGenieTeam_importWithUser
--- PASS: TestAccOpsGenieTeam_importWithUser (6.43s)
=== RUN   TestAccOpsGenieTeam_importWithUserComplete
--- PASS: TestAccOpsGenieTeam_importWithUserComplete (7.50s)
=== RUN   TestAccOpsGenieUser_importBasic
--- PASS: TestAccOpsGenieUser_importBasic (3.24s)
=== RUN   TestAccOpsGenieUser_importComplete
--- PASS: TestAccOpsGenieUser_importComplete (3.60s)
=== RUN   TestAccOpsGenieTeamName_validation
--- PASS: TestAccOpsGenieTeamName_validation (0.00s)
=== RUN   TestAccOpsGenieTeamRole_validation
--- PASS: TestAccOpsGenieTeamRole_validation (0.00s)
=== RUN   TestAccOpsGenieTeam_basic
--- PASS: TestAccOpsGenieTeam_basic (2.78s)
=== RUN   TestAccOpsGenieTeam_withUser
--- PASS: TestAccOpsGenieTeam_withUser (5.62s)
=== RUN   TestAccOpsGenieTeam_withUserComplete
--- PASS: TestAccOpsGenieTeam_withUserComplete (5.62s)
=== RUN   TestAccOpsGenieUserUsername_validation
--- PASS: TestAccOpsGenieUserUsername_validation (0.00s)
=== RUN   TestAccOpsGenieUserFullName_validation
--- PASS: TestAccOpsGenieUserFullName_validation (0.00s)
=== RUN   TestAccOpsGenieUserRole_validation
--- PASS: TestAccOpsGenieUserRole_validation (0.00s)
=== RUN   TestAccOpsGenieUser_basic
--- PASS: TestAccOpsGenieUser_basic (3.08s)
=== RUN   TestAccOpsGenieUser_complete
--- PASS: TestAccOpsGenieUser_complete (3.03s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/opsgenie	52.762s
```

File: builtin/providers/opsgenie/resource_opsgenie_team.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+	}
+
+	return nil
+}
+
+func resourceOpsGenieTeamDelete(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[INFO] Deleting OpsGenie team '%s'", d.Get("name").(string))
+	client := meta.(*OpsGenieClient).teams
+
+	deleteRequest := team.DeleteTeamRequest{
```
Comment (stack72) [inline builtin/providers/opsgenie/resource_opsgenie_team.go]: Will this be allowed if there are members attached to the Team? e.g. AWS doesn't allow us to delete an iam_role if policies are attached to it

File: builtin/providers/opsgenie/resource_opsgenie_team_test.go [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+resource "opsgenie_user" "test" {
+  username  = "acctest-%d@example.tld"
+  full_name = "Acceptance Test User"
+  role      = "User"
+}
+
+resource "opsgenie_team" "test" {
+  name  = "acctest%d"
+  member {
+    username = "${opsgenie_user.test.username}"
```
Comment (stack72) [inline builtin/providers/opsgenie/resource_opsgenie_team_test.go]: I'd like to see a test that adds multiple users to a team

File: builtin/providers/opsgenie/provider.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+		},
+
+		ConfigureFunc: providerConfigure,
+	}
+}
+
+func (c *Config) validate() error {
+	var err *multierror.Error
+
+	if c.ApiKey == "" {
```
Comment (tombuildsstuff) [inline builtin/providers/opsgenie/provider.go]: Good spot

File: builtin/providers/opsgenie/resource_opsgenie_team_test.go (hunk 3 above)
Comment (tombuildsstuff) [inline builtin/providers/opsgenie/resource_opsgenie_team_test.go]: done

**Comment by tombuildsstuff** [comment]
Tests still pass:

```
$ TF_ACC=1 go test ./builtin/providers/opsgenie -v -run TestAcc -timeout 120m
=== RUN   TestAccDataSourceOpsGenieUser_Basic
--- PASS: TestAccDataSourceOpsGenieUser_Basic (9.65s)
=== RUN   TestAccOpsGenieTeam_importBasic
--- PASS: TestAccOpsGenieTeam_importBasic (5.01s)
=== RUN   TestAccOpsGenieTeam_importWithUser
--- PASS: TestAccOpsGenieTeam_importWithUser (9.59s)
=== RUN   TestAccOpsGenieTeam_importWithUserComplete
--- PASS: TestAccOpsGenieTeam_importWithUserComplete (7.65s)
=== RUN   TestAccOpsGenieUser_importBasic
--- PASS: TestAccOpsGenieUser_importBasic (4.27s)
=== RUN   TestAccOpsGenieUser_importComplete
--- PASS: TestAccOpsGenieUser_importComplete (4.71s)
=== RUN   TestAccOpsGenieTeamName_validation
--- PASS: TestAccOpsGenieTeamName_validation (0.00s)
=== RUN   TestAccOpsGenieTeamRole_validation
--- PASS: TestAccOpsGenieTeamRole_validation (0.00s)
=== RUN   TestAccOpsGenieTeam_basic
--- PASS: TestAccOpsGenieTeam_basic (3.58s)
=== RUN   TestAccOpsGenieTeam_withUser
--- PASS: TestAccOpsGenieTeam_withUser (7.82s)
=== RUN   TestAccOpsGenieTeam_withUserComplete
--- PASS: TestAccOpsGenieTeam_withUserComplete (6.95s)
=== RUN   TestAccOpsGenieTeam_withMultipleUsers
--- PASS: TestAccOpsGenieTeam_withMultipleUsers (7.50s)
=== RUN   TestAccOpsGenieUserUsername_validation
--- PASS: TestAccOpsGenieUserUsername_validation (0.00s)
=== RUN   TestAccOpsGenieUserFullName_validation
--- PASS: TestAccOpsGenieUserFullName_validation (0.00s)
=== RUN   TestAccOpsGenieUserRole_validation
--- PASS: TestAccOpsGenieUserRole_validation (0.00s)
=== RUN   TestAccOpsGenieUser_basic
--- PASS: TestAccOpsGenieUser_basic (3.65s)
=== RUN   TestAccOpsGenieUser_complete
--- PASS: TestAccOpsGenieUser_complete (3.48s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/opsgenie	73.862s
```

File: builtin/providers/opsgenie/resource_opsgenie_team.go (hunk 2 above)
Comment (tombuildsstuff) [inline builtin/providers/opsgenie/resource_opsgenie_team.go]: I believe so, but I haven't tried with a whole chain (i.e. Schedules etc) - let me take a look

**Comment by stack72** [comment]
This is looking great now - thanks for making the changes I requested :)

```
% make testacc TEST=./builtin/providers/opsgenie                                                                              ✭
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/01/05 21:23:23 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/opsgenie -v  -timeout 120m
=== RUN   TestAccDataSourceOpsGenieUser_Basic
--- PASS: TestAccDataSourceOpsGenieUser_Basic (6.39s)
=== RUN   TestAccOpsGenieTeam_importBasic
--- PASS: TestAccOpsGenieTeam_importBasic (4.00s)
=== RUN   TestAccOpsGenieTeam_importWithUser
--- PASS: TestAccOpsGenieTeam_importWithUser (6.64s)
=== RUN   TestAccOpsGenieTeam_importWithUserComplete
--- PASS: TestAccOpsGenieTeam_importWithUserComplete (7.10s)
=== RUN   TestAccOpsGenieUser_importBasic
--- PASS: TestAccOpsGenieUser_importBasic (3.99s)
=== RUN   TestAccOpsGenieUser_importComplete
--- PASS: TestAccOpsGenieUser_importComplete (3.78s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccOpsGenieTeamName_validation
--- PASS: TestAccOpsGenieTeamName_validation (0.00s)
=== RUN   TestAccOpsGenieTeamRole_validation
--- PASS: TestAccOpsGenieTeamRole_validation (0.00s)
=== RUN   TestAccOpsGenieTeam_basic
--- PASS: TestAccOpsGenieTeam_basic (3.09s)
=== RUN   TestAccOpsGenieTeam_withUser
--- PASS: TestAccOpsGenieTeam_withUser (7.03s)
=== RUN   TestAccOpsGenieTeam_withUserComplete
--- PASS: TestAccOpsGenieTeam_withUserComplete (6.08s)
=== RUN   TestAccOpsGenieTeam_withMultipleUsers
--- PASS: TestAccOpsGenieTeam_withMultipleUsers (7.04s)
=== RUN   TestAccOpsGenieUserUsername_validation
--- PASS: TestAccOpsGenieUserUsername_validation (0.00s)
=== RUN   TestAccOpsGenieUserFullName_validation
--- PASS: TestAccOpsGenieUserFullName_validation (0.00s)
=== RUN   TestAccOpsGenieUserRole_validation
--- PASS: TestAccOpsGenieUserRole_validation (0.00s)
=== RUN   TestAccOpsGenieUser_basic
--- PASS: TestAccOpsGenieUser_basic (3.40s)
=== RUN   TestAccOpsGenieUser_complete
--- PASS: TestAccOpsGenieUser_complete (4.18s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/opsgenie	62.711s
```

**Pull request merged**